e2e/imgbuild/imgbuild.go

```diff
@@ -154,6 +159,511 @@ func badPath(t *testing.T) {
+					{
+						Src: tmpfile.Name(),
+						Dst: "StageOneCopy.txt",
+					},
+				},
+			},
+		},
+		// Complex copy of files from stage one and two to stage three, then final copy from three to final stage
+		{"FileCopyComplex", false, true,
+			[]e2e.DefFileDetails{
+				{
+					Bootstrap: "docker",
+					From:      "alpine:latest",
+					Stage:     "one",
+					Files: []e2e.FilePair{
+						{
+							Src: tmpfile.Name(),
+							Dst: "StageOne2.txt",
+						},
+						{
+							Src: tmpfile.Name(),
+							Dst: "StageOne.txt",
+						},
+					},
+				},
+				{
+					Bootstrap: "docker",
+					From:      "alpine:latest",
+					Stage:     "two",
+					Files: []e2e.FilePair{
+						{
+							Src: tmpfile.Name(),
+							Dst: "StageTwo2.txt",
+						},
+						{
+							Src: tmpfile.Name(),
+							Dst: "StageTwo.txt",
+						},
+					},
+				},
+				{
+					Bootstrap: "docker",
+					From:      "alpine:latest",
+					Stage:     "three",
+					FilesFrom: []e2e.FileSection{
+						{
+							Stage: "one",
+							Files: []e2e.FilePair{
+								{
+									Src: "StageOne2.txt",
+									Dst: "StageOneCopy2.txt",
+								},
+								{
+									Src: "StageOne.txt",
+									Dst: "StageOneCopy.txt",
+								},
+							}},
+						{
+							Stage: "two",
+							Files: []e2e.FilePair{
+								{
+									Src: "StageTwo2.txt",
+									Dst: "StageTwoCopy2.txt",
+								},
+								{
+									Src: "StageTwo.txt",
+									Dst: "StageTwoCopy.txt",
+								},
+							},
+						}},
+				},
+				{
+					Bootstrap: "docker",
+					From:      "alpine:latest",
+					FilesFrom: []e2e.FileSection{
+						{
+							Stage: "three",
+							Files: []e2e.FilePair{
+								{
+									Src: "StageOneCopy2.txt",
+									Dst: "StageOneCopyFinal2.txt",
+								},
+								{
+									Src: "StageOneCopy.txt",
+									Dst: "StageOneCopyFinal.txt",
+								},
+								{
+									Src: "StageTwoCopy2.txt",
+									Dst: "StageTwoCopyFinal2.txt",
+								},
+								{
+									Src: "StageTwoCopy.txt",
+									Dst: "StageTwoCopyFinal.txt",
+								},
+							}}},
+				},
+			},
+			e2e.DefFileDetails{
+				Files: []e2e.FilePair{
+					{
+						Src: tmpfile.Name(),
+						Dst: "StageOneCopyFinal2.txt",
+					},
+					{
+						Src: tmpfile.Name(),
+						Dst: "StageOneCopyFinal.txt",
+					},
+					{
+						Src: tmpfile.Name(),
+						Dst: "StageTwoCopyFinal2.txt",
+					},
+					{
+						Src: tmpfile.Name(),
+						Dst: "StageTwoCopyFinal.txt",
+					},
+				},
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, test.WithPrivilege(func(t *testing.T) {
+
+			defFile := e2e.PrepareMultiStageDefFile(tt.dfd)
+			defer os.Remove(defFile)
+
+			opts := e2e.BuildOpts{
+				Sandbox: tt.sandbox,
+			}
+
+			imagePath := path.Join(testenv.TestDir, "container")
+			defer os.RemoveAll(imagePath)
+
+			if b, err := e2e.ImageBuild(testenv.CmdPath, opts, imagePath, defFile); err != nil {
+				t.Log(string(b))
+				t.Fatalf("unexpected failure: %v", err)
```

Comment:
All of your `t.Fatalf` calls use `%v` for the format, but I think you can use `%s` instead.

Reply:
I wonder what the new error handling feature will want us to use? Maybe just leave this for now and we can raise the issue globally later :)